spec/app/ics-031-crosschain-queries/README.md

```diff
@@ -0,0 +1,388 @@
+- The `localTimeoutTimestamp` field specifies a timestamp limit at the Querying Chain after which a query is considered to have failed and a timeout result should be returned to the original caller.
+- The `queryHeight` field is the height at which the relayer must query the Queried Chain
+- The `clientId` field identifies the Queried Chain.
+- The `bounty` field is a bounty that is given to the relayer for participating in the query.
+
+The Cross-chain Querying module stores query results to allow query callers to asynchronously retrieve them. 
+In this context, this ICS defines the `QueryResult` type as follows:
+
+```typescript
+enum QueryResult {
+  SUCCESS,
+  FAILURE,
+  TIMEOUT,
+}
+```
+- A query that returns a value is marked as `SUCCESS`. This means that the query has been executed at the Queried Chain and there was a value associated to the queried path at the requested height.
+- A query that is executed but does not return a value is marked as `FAILURE`. This means that the query has been executed at the Queried Chain, but there was no value associated to the queried path at the requested height.
+- A query that timed out before a result is committed at the Querying Chain is marked as `TIMEOUT`.
+
+A CrossChainQueryResult is a particular interface used to represent query results.
+
+```typescript
+interface CrossChainQueryResult struct {
+    id: Identifier
+    result: QueryResult
+    data: []byte
+}
+```
+
+- The `id` field uniquely identifies the query at the Querying Chain.
+- The `result` field indicates whether the query was correctly executed at the Queried Chain and if the queried path exists.
+- The `data` field is an opaque bytestring that contains the value associated with the queried path in case `result = SUCCESS`.
+
+### Store paths
+
+#### Query path
+
+The query path is a private path that stores the state of ongoing cross-chain queries.
+
+```typescript
+function queryPath(id: Identifier): Path {
+    return "queries/{id}"
+}
+```
+#### Result query path
+
+The result query path is a private path that stores the result of completed queries.
+
+```typescript
+function resultQueryPath(id: Identifier): Path {
+    return "queriesresult/{id}"
+}
+```
+
+### Helper functions
+
+The Querying Chain MUST implement a function `generateQueryIdentifier`, which generates a unique query identifier:
+
+```typescript
+function generateQueryIdentifier = () -> Identifier
+```
+
+### Sub-protocols
+
+#### Query lifecycle
+
+1) When the Querying Chain receives a query request, it calls `CrossChainQueryRequest` of the Cross-chain Querying module. This function generates a unique identifier for the query, stores it in its `privateStore` and emits a `sendQuery` event. Query requests can be submitted by other IBC modules as transactions to the Querying Chain or simply executed as part of the `BeginBlock` and `EndBlock` logic.
```

Comment:
> Query requests can be submitted by other IBC modules

Why only IBC modules?

Reply:
I think I discussed this with Aditya. I believe the conclusion was that in most of the cases, cross-chain queries will be issued by IBC modules, as end users would directly query the queried chain. At least this is what I remember, but maybe @AdityaSripal can elaborate a bit more. In any case, I think that even if that's the common case, we should not constraint the specification, so I will fix the text.